subcommand/injector/command.go

```diff
@@ -316,29 +317,53 @@ func getAdminAPIVersion(ctx context.Context, clientset *kubernetes.Clientset) (s
 	return adminAPIVersion, err
 }
 
-func (c *Command) certWatcher(ctx context.Context, ch <-chan cert.Bundle, clientset *kubernetes.Clientset, leaderElector leader.Elector, adminAPIVersion string, log hclog.Logger) {
+func (c *Command) certWatcher(ctx context.Context, ch <-chan cert.Bundle, clientset *kubernetes.Clientset, leaderElector leader.Elector, namespace string, log hclog.Logger) {
 	var bundle cert.Bundle
+	webhooksCache, webhooksWatcher, err := watchWebhooks(ctx, clientset, namespace)
+	if err != nil {
+		log.Error("Error watching webhooks", "error", err)
+		panic(err)
+	}
+	tryImmediately := false
 
 	for {
-		select {
-		case bundle = <-ch:
-			log.Info("Updated certificate bundle received. Updating certs...")
-			// Bundle is updated, set it up
-
-		case <-time.After(1 * time.Second):
-			// This forces the mutating webhook config to remain updated
-			// fairly quickly. This is a jank way to do this and we should
-			// look to improve it in the future. Since we use Patch requests
-			// it is pretty cheap to do, though.
+		if tryImmediately {
+			// avoid tight loop on failure
+			time.Sleep(1 * time.Second)
```

Comment:
We could also wrap all of the logic that's after the `select` within this for loop to its own func which sends the error to an `errCh` that's also hooked into the select here (and we can sleep for 1s in there instead). It has the benefit of getting rid of the additional nesting and various `continue` paths that exists here.

Reply:
:+1:

Reply:
Re: `errCh`, it didn't make a lot of sense to add a separate channel for that, since we don't want to execute the update function as a goroutine (we could get into some race conditions).

But, it does make sense to move it to a separate function and simplify the backoff logic a little, so I did that. Thanks for the suggestion.